reflect-lint/src/main/java/dagger/reflect/lint/WrongRetentionDetector.java

```diff
@@ -0,0 +1,189 @@
+            actualRetention),
+        LintFix.create()
+            .name("Replace with: `@Retention(RUNTIME)`")
+            .replace()
+            .range(context.getLocation(annotationValue))
+            .with(isKotlin ? FIX_RETENTION_TYPE_KOTLIN : FIX_RETENTION_TYPE_JAVA)
+            .reformat(true)
+            .shortenNames()
+            .build());
+  }
+
+  @NotNull
+  private static String getRetentionPolicy(
+      @NotNull JavaContext context, boolean isKotlin, @NotNull UAnnotation retentationAnnotation) {
+    final UExpression annotationValue = UastLintUtils.getAnnotationValue(retentationAnnotation);
+    final String retentionPolicyQualifiedName =
+        isKotlin
+            ? getQualifiedNameForValueKotlin(context, annotationValue)
+            : getQualifiedNameForValueJava(context, annotationValue);
+    final String retentionPolicy = getRetentionPolicyForQualifiedName(retentionPolicyQualifiedName);
+    if (retentionPolicy != null) {
+      return retentionPolicy;
+    }
+    throw new IllegalStateException("RetentionPolicy must not be null if @Retention is present");
+  }
+
+  @NotNull
+  private static String getQualifiedNameForValueKotlin(
+      @NotNull JavaContext context, @Nullable UExpression annotationValue) {
+    final Object evaluatedAnnotationValue = ConstantEvaluator.evaluate(context, annotationValue);
+    if (evaluatedAnnotationValue instanceof kotlin.Pair) {
+      final kotlin.Pair value = (kotlin.Pair) evaluatedAnnotationValue;
+      final String qualifiedName = (value.getFirst() + "." + value.getSecond());
+      return qualifiedName.replace("/", ".");
+    }
+    throw new IllegalStateException("RetentionPolicy must not be null if @Retention is present");
+  }
+
+  @NotNull
+  private static String getQualifiedNameForValueJava(
+      @NotNull JavaContext context, @Nullable UExpression annotationValue) {
+    final Object evaluatedAnnotationValue = ConstantEvaluator.evaluate(context, annotationValue);
+    if (evaluatedAnnotationValue instanceof PsiEnumConstant) {
+      return UastLintUtils.getQualifiedName((PsiEnumConstant) evaluatedAnnotationValue);
+    }
+    throw new IllegalStateException("RetentionPolicy must not be null if @Retention is present");
+  }
+
+  @Nullable
+  private static String getRetentionPolicyForQualifiedName(@NotNull String retentionPolicy) {
+    // Values are same for Kotlin and Java
+    for (RetentionPolicy policy : RetentionPolicy.values()) {
+      final String javaQualifiedName = CLASS_JAVA_RETENTION_POLICY + "." + policy.name();
+      final String kotlinQualifiedName = CLASS_KOTLIN_RETENTION_POLICY + "." + policy.name();
+      if (javaQualifiedName.equals(retentionPolicy)
+          || kotlinQualifiedName.equals(retentionPolicy)) {
+        return policy.name();
+      }
+    }
+    return null;
+  }
+
+  public static final Issue ISSUE_MISSING_RETENTION =
```

Comment:
I don't think it's useful to have these as two checks instead of one. When no annotation is specified there is a default retention (of class) so it's really the same situation as having the wrong value. Even if that weren't true, I don't think the distinction is that useful.

Reply:
Good point. I removed the separate issue but kept the logic the same such that the error messages are still different.